refactor: Replace RecursiveMutex with Mutex in timedata.cpp

hebasto opened the pull request:
Only `GetTimeOffset()` and `AddTimeData()` functions lock this mutex. They do not call itself recursively, and do not call each other either directly or indirectly. Therefore, the `g_timeoffset_mutex` could be a non-recursive mutex.

Related to #19180.

hebasto committed:
refactor: Replace RecursiveMutex with Mutex in timedata.cpp

MarcoFalke commented:
Concept ACK. (nit: Could add a new commit to clang-format the whole file while touching it anyway?)

hebasto committed:
refactor: Fix formatting of timedata.cpp

hebasto commented:
Updated c2410ceb844a443caf6dd8c6df976b9e24724d06 -> cc5c0d2299b09c58cd9962ca5075ffa53f2633c0 ([pr19189.01](https://github.com/hebasto/bitcoin/commits/pr19189.01) -> [pr19189.02](https://github.com/hebasto/bitcoin/commits/pr19189.02), [diff](https://github.com/hebasto/bitcoin/compare/pr19189.01..pr19189.02)):

- addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/19189#issuecomment-640197243):
> nit: Could add a new commit to clang-format the whole file while touching it anyway?

MarcoFalke commented:
ACK cc5c0d2299b09c58cd9962ca5075ffa53f2633c0 , checked the second commit with  --word-diff-regex=. --ignore-all-space -U0 🦉

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK cc5c0d2299b09c58cd9962ca5075ffa53f2633c0 , checked the second commit with  --word-diff-regex=. --ignore-all-space -U0 🦉
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUg9ywv/RuHDTDtPin6ga49t7Vslgym8iyJlbJNII/rcUL4iqxTbocvKoIpaJS4O
URw7HGDe8sr5pBCEuwtM1lKgOxR9JQ2grVLisimxl99rQdLUd8ryCfS5KfAj4Ta+
R5P1IMKO7/oSlovhcnWCoF7IXFcEDbSlYIz+OqYpDYBM6VtadUHTxvy37Tq/nLxO
N+V2CZa2f09toH35OVv019BdSOQ/GiaiAAyMHY4Fwzjgyl7hLTa2DUKuGY/9Myt1
goCN/C+aqtk/O0eNmF87Tq9yARaoba70L9zhNafGNgDB7XFnPY72xrJueCaY/+lm
G2m24ZUVmGbb8NolwXJ6/57J0nysNu/aDZFsDXTpweb3px/uIsOtuivD6zXkMo7f
gvAJ274hCz6DJTXWkyDPJOEk99SFbn6aQaayIBLhdoo5HxihmNNUKhuBnQl0pSPw
5ZsVLD8WQTIPhAeblnm3U6E3ytF2UrZK0yq0yuNWwf8C0Z5YJceUKvYHnM09o0KC
KTFf3Fc3
=r9Js
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `bfef319e8de4b4be6ce092abda9bc310590248e63368de1b85795fcaab7291cf  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108bfef319e8de4b4be6ce092abda9bc310590248e63368de1b85795fcaab7291cff010ca3f2e65fd33d888375ca0d4adfb24ef08fff01007a0adeda1afcc41c4b99848505138df08f1045edcd256f008da844b45bc8b855a0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff010649f819bdd90898fa43f336fc361814208f1045edcd257f008865c085e93f3a91b0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff010b8e1d6460e2b1aa44d83ad379a1635a308f1045edcd256f008d387aa042ea050450083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010c9056d8757d331b572b102e941beedb908f1045edcd256f008e8519bbbeb115b370083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

vasild commented on src/timedata.cpp:
```diff
@@ -76,30 +75,24 @@ void AddTimeData(const CNetAddr& ip, int64_t nOffsetSample)
-                    if (nOffset != 0 && abs64(nOffset) < 5 * 60)
-                        fMatch = true;
+                for (const int64_t nOffset : vSorted) {
+                    if (nOffset != 0 && abs64(nOffset) < 5 * 60) fMatch = true;
```
I know clang-format did it due to `AllowShortIfStatementsOnASingleLine: true` and so there is nothing to do in this PR, but I just want to mention that putting the `if` condition and body on one line like `if (A) B;`
* makes life in gdb more difficult (if one wants to set a breakpoint on B)
* makes backtraces less usefull (if it crashed on line 92, did it crash in `A` or in `B`?)
* skews line based code coverage (if just `A` is executed and it was `false` then the code coverage may show the entire line 92 covered, giving the false impression that `B` was also executed).

hebasto replied:
Lean to agree with you. Just want to preserve ACKs :)

MarcoFalke replied:
We only allow trivial statements in single line ifs, so it should always be clear whether it crashed in A or B.

Also, coverage reports reports generally include branch coverage.

vasild commented on src/timedata.cpp:
```diff
@@ -15,9 +15,8 @@
-static RecursiveMutex cs_nTimeOffset;
-static int64_t nTimeOffset GUARDED_BY(cs_nTimeOffset) = 0;
+static Mutex g_timeoffset_mutex;
+static int64_t nTimeOffset GUARDED_BY(g_timeoffset_mutex) = 0;
```
nit: `s/nTimeOffset/g_time_offset/`, feel free to ignore. I just wonder - you changed the name of one global variable why not the other too?

hebasto replied:
I hope this commit is mutex related only :)

vasild reviewed (approved):
ACK cc5c0d22 verified that recursion is not happening

hebasto commented:
@sipa Mind reviewing this PR?
